Geometry_Engine/Query/IsVertical.cs

```diff
@@ -0,0 +1,54 @@
+/*
+ * This file is part of the Buildings and Habitats object Model (BHoM)
+ * Copyright (c) 2015 - 2020, the respective contributors. All rights reserved.
+ *
+ * Each contributor holds copyright over their respective contributions.
+ * The project versioning (Git) records all such contribution source information.
+ *                                           
+ *                                                                              
+ * The BHoM is free software: you can redistribute it and/or modify         
+ * it under the terms of the GNU Lesser General Public License as published by  
+ * the Free Software Foundation, either version 3.0 of the License, or          
+ * (at your option) any later version.                                          
+ *                                                                              
+ * The BHoM is distributed in the hope that it will be useful,              
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of               
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                 
+ * GNU Lesser General Public License for more details.                          
+ *                                                                            
+ * You should have received a copy of the GNU Lesser General Public License     
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using BH.oM.Geometry;
+using System;
+
+namespace BH.Engine.Geometry
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Private Methods                           ****/
+        /***************************************************/
+
+        public static bool IsVertical(this Line line)
+        {
+            return IsVertical(line.Start, line.End);
+        }
+
+        /***************************************************/
+        /**** Private Methods                           ****/
+        /***************************************************/
+
+        private static bool IsVertical(Point p1, Point p2)
+        {
+            double dx = p1.X - p2.X;
+            double dy = p1.Y - p2.Y;
+
+            return Math.Sqrt(dx * dx + dy * dy) < 0.0001;
```

Comment:
Is this a tolerance `0.0001` or part of a pre-determined calculation? If it's a tolerance, I'd advise we expose it as an input and give it the default from `BH.oM.Geometry.Tolerance`

Reply:
This is copied code (from the BHoM), and it is defined in the [documentation](https://github.com/BHoM/documentation/wiki/BHoM-Structural-Conventions). Or do we want to change it @IsakNaslundBh ?

Reply:
The documentation says:
 > A bar is vertical if it's projected length to the horizontal plane is less than 0.001

Whereas your code says `< 0.0001` (an extra 0) - if it's to match the documentation then one of them is wrong (or I'm looking at the wrong bit of documentation?) so might need to change one of them.

If this is a structural convention then that would be ok for when it was in the Structure_Engine, but if this is now coming into the more common Geometry_Engine I think exposing that parameter would be more suitable? @IsakNaslundBh - thoughts?

Reply:
Would guess that it's a typo in the documentation since it also says 0.1mm directly contradicting it self, but does match the code. But would also like some thoughts @IsakNaslundBh

Reply:
> Is this a tolerance 0.0001 or part of a pre-determined calculation? If it's a tolerance, I'd advise we expose it as an input and give it the default from BH.oM.Geometry.Tolerance

Documentation is wrong. Will fix. Good spot, thanks! 👍 

> If this is a structural convention then that would be ok for when it was in the Structure_Engine, but if this is now coming into the more common Geometry_Engine I think exposing that parameter would be more suitable? @IsakNaslundBh - thoughts?

There are multiple conventions for this, and it varies across software packages etc. The important thing is to have one singly defined convention somewhere. I am not particularly attached to the current convention we have and be fine to change it to something else if something is suggested, but as I said, we need it defined somewhere, and only once.

The important part is that this _needs_ to be consistent, _always_. That is the reason for the hard coded value, rather than having it exposed. We could potentially solve this by exposing the value in geometry engine and hard coding it in the physical and structure engine, but that would mean typing it out twice, and hence have the potential of them getting out of sync, which is something I really want to avoid.

IDK if this is enough for you @FraserGreenroyd ? If not we might have to think of another solution to this whole PR.

Reply:
> IDK if this is enough for you @FraserGreenroyd ? If not we might have to think of another solution to this whole PR.

I'm think this PR solves the problem best for now so I don't think we should spend more time on another solution in itself.

I am happy with a 1mm tolerance being the standard and hard-coding it (at least for now) if we're happy it meets the criteria required given its new location (Geometry_Engine rather than Structure_Engine). To avoid hard-coding in Physical/Structure engines though, we could do a tolerance input which has a default input of `0.001` but this of course hides that tolerance from the code side on Physical/Structural (though this is no different to the current implementation).

Ultimately, I'm not too concerned as I've raised my query and I feel this satisfactorily answers it for now, so if you're happy with this implementation for now @IsakNaslundBh then I'm happy to support it :smile: